src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+        /// <param name="end">The ending (inclusive) <see cref="SequencePosition"/> of the slice</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start, SequencePosition end)
+        {
+            BoundsCheck(_bufferEnd, end);
+            BoundsCheck(end, start);
+
+            return SliceImpl(start, end);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start)
+        {
+            BoundsCheck(_bufferEnd, start);
+
+            return SliceImpl(start, _bufferEnd);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="start">The start index at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(long start)
+        {
+            if (start == 0) return this;
+
+            SequencePosition begin = Seek(_bufferStart, _bufferEnd, start, false);
+            return SliceImpl(begin, _bufferEnd);
+        }
+
+        /// <inheritdoc />
+        public override string ToString() => string.Format("System.Buffers.ReadOnlyBuffer<{0}>[{1}]", typeof(T).Name, Length);
+
+        /// <summary>
+        /// Returns an enumerator over the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public Enumerator GetEnumerator() => new Enumerator(this);
+
+        /// <summary>
+        /// Returns new <see cref="SequencePosition"/> that is offset by <paramref name="offset"/> starting with <paramref name="origin"/>
+        /// </summary>
+        public SequencePosition GetPosition(SequencePosition origin, long offset)
+        {
+            if (offset < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.offset);
+            }
+            return Seek(origin, _bufferEnd, offset, false);
+        }
+
+        /// <summary>
+        /// Tries to retrieve next segment after <paramref name="position"/> and return it's contents in <paramref name="data"/>.
+        /// Returns <code>false</code> if end of <see cref="ReadOnlyBuffer{T}"/> was reached otherwise <code>false</code>.
+        /// Sets <paramref name="position"/> to the beginning of next segment is <paramref name="advance"/> is set to <code>true</code>.
+        /// </summary>
+        public bool TryGet(ref SequencePosition position, out ReadOnlyMemory<T> data, bool advance = true)
+        {
+            bool result = TryGetBuffer(position, End, out data, out SequencePosition next);
```

Comment:
Can we change TryGetBuffer to take in the ref SequencePosition and advance bool, so we can just set it there rather than having to assign next to position here?

Reply:
It would increase nesting a lot.